clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -207,6 +226,232 @@ public NetworkClientDelegate.UnsentRequest toUnsentRequest() {
+                    groupState.groupId,
+                    groupState.groupInstanceId.orElse(null),
+                    groupState.generation);
+            unsentOffsetCommits.add(request);
+            return request.future();
+        }
+
+        /**
+         *  We want to avoid duplication when sending an {@link OffsetFetchRequest}. The following checks are done:
+         *  <li>1. dedupe against unsents: if a duplicated request was previously made, we chain the future</>
+         *  <li>2. dedupe against incompleted: If a duplicated request was sent but hasn't gotten a results, we chain
+         *  the future.</>
+         *
+         *  <p>If the request is new, we chain a call back to remove itself from the {@code inflightOffsetFetches}
+         *  upon completion.</>
+         */
+        private CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> addOffsetFetchRequest(final UnsentOffsetFetchRequest request) {
+            Optional<UnsentOffsetFetchRequest> dupe =
+                    unsentOffsetFetches.stream().filter(r -> r.sameRequest(request)).findAny();
+            Optional<UnsentOffsetFetchRequest> inflight =
+                    inflightOffsetFetches.stream().filter(r -> r.sameRequest(request)).findAny();
+
+            if (dupe.isPresent() || inflight.isPresent()) {
+                log.info("Duplicated OffsetFetchRequest: " + request.requestedPartitions);
+                dupe.orElseGet(() -> inflight.get()).chainFuture(request.future);
+            } else {
+                // remove the request from the outbound buffer: inflightOffsetFetches
+                request.future.whenComplete((r, t) -> {
+                    if (!inflightOffsetFetches.remove(request)) {
+                        log.info("unable to remove request from the outbound buffer:" + request);
+                    }
+                });
+                this.unsentOffsetFetches.add(request);
+            }
+            return request.future;
+        }
+
+        private CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> addOffsetFetchRequest(final Set<TopicPartition> partitions) {
+            UnsentOffsetFetchRequest request = new UnsentOffsetFetchRequest(
+                    partitions,
+                    groupState.generation,
+                    retryBackoffMs);
+            return addOffsetFetchRequest(request);
+        }
+
+        public List<NetworkClientDelegate.UnsentRequest> drain(final long currentTimeMs) {
+            List<NetworkClientDelegate.UnsentRequest> unsent = new ArrayList<>();
+            unsent.addAll(unsentOffsetCommits.stream()
+                    .map(UnsentOffsetCommitRequest::toUnsentRequest)
+                    .collect(Collectors.toList()));
+            List<UnsentOffsetFetchRequest> sendables = unsentOffsetFetches.stream()
+                    .filter(r -> r.canSendRequest(currentTimeMs))
+                    .collect(Collectors.toList());
+            inflightOffsetFetches.addAll(sendables);
```

Comment:
This seems not right to me: if we add the inflight request to sendables again and return in `drain`, then they will be re-sent by the network client delegate?

Reply:
The `inflightOffsetFetches` is just a temporary queue for bookkeeping purposes. So the drain does two things:
1. Return the requests that need to be send
2. Put these requests into the `inflightOffsetFetches` queue for tracking.  So that we know what is currently inflight.

As you can see, there's no direct interaction with the NetworkClient in the RequestManager, so no request is actually making to the networkClient.  I think the requirement is that the background thread would pool these request managers and queue up the poll results into NetworkClient directly.